### DISCUSSION TITLE
Google Search Console and Next.js caching - "_rsc=" issue

### DISCUSSION THREAD

**Discussion by Antonio112009** [vercel/next.js#61850] [Help]
### Summary

I have noticed an increasing number of errors in my Google Search Console related to caching in Next.js. Specifically, the error is 'Alternative page with proper canonical tag' and 98% of occurrences are related to the caching issue. I would like to know how to prevent Google from indexing pages with a URL query '_rsc='. 

I tried excluding these pages from indexing by adding them to the robots.txt file, but as they are necessary pages, that approach did not work. Can someone please advise me on what I should do?

<img width="879" alt="image" src="https://github.com/vercel/next.js/assets/29264726/eb371e1b-c6b9-4af1-9599-8f4194cf139f">

**Comment by nykula** [comment]
I'd also appreciate someone familiar with Next.js internals hypothesizing why Google considers the _rsc AJAX calls worthy of being compared to their equivalent public URLs despite _rsc never explicitly appearing in anyone's address bar, and whether anything can or should be done about that by developers who use the Next.js App Router and Link component or by developers of Next.js itself.

**Comment by Refrezsh** [comment]
These pages do not present any new content. The solution for myself was to add a line in Robots.txt.

`Disallow: /*?_rsc=*`

**Reply by Johrten** [reply]
Is this still the solution you have implemented? I added this but Google just moved to "Indexed, though blocked by robots.txt"

**Comment by Midnight-Knight** [comment]
I came across a similar issue regarding Google and Google Search Console. For some reason, Google is massively crawling pages with "_rsc" parameters in the URL, which leads to errors in identifying canonical pages, despite the fact that they are set for each one. It can also cause incorrect indexing of pages. As a result, our site has stopped being indexed in Google, although everything is fine in other search engines. 

We tried to solve this problem using a file ```robots.txt``` but it had no effect. 

We also tried to fix this by checking the URL parameters and using a redirect in ```middleware.ts```, but the ```NextRequest``` does not have such a parameter, since Next removes it from the fields. For example, if the URL states ```"http://localhost:3000 /?__rsc=1ld0r"```, then in middleware the ```request: Next Request``` parameters are as follows:
```
{
  href: 'http://localhost:3000/',
  origin: 'http://localhost:3000',
  protocol: 'http:',
  username: ''password: '',
  host: 'localhost:3000',
  hostname: 'localhost',
  port: '3000',
  pathname: '/',
  search: '',
  searchParams: URLSearchParams {  },
  hash: ''
}
```
Thus, Next.for some reason, js clears these parameters. Perhaps someone has encountered a similar situation?

**Reply by dennzimm** [reply]
Have you already found a solution? Or how did you proceed with this?

**Reply by Midnight-Knight** [reply]
Hi, at the moment, it has not been possible to completely fix the problem, but there is a feeling that we are moving in the right direction. It turned out that the ```_rsc``` query parameter is not available in any way for manipulation in ```middleware.ts```, however, you can work with it in ```next.config.js``` . The following code allows you to configure traffic redirection if the ```_rsc``` query parameter is present in the URL. 
```
module.exports = {
  async redirects() {
    return [
      {
        source: '/',
        destination: '/',
        has: [
          {
            type: 'query',
            key: '_rsc',
          }
        ],
        permanent: true,
      },
      {
        source: '/:path*',
        destination: '/:path*',
        has: [
          {
            type: 'query',
            key: '_rsc',
          }
        ],
        permanent: true,
      }
    ]
  },
  ...
}
```
As a result, from 1300+ original errors related to ```_rsc``` only 402 remain at the moment. Google started indexing some pages, but they are still not showing up in the search results, so the problem hasn't been fully resolved yet.

The version is next.js - 15.1.4

**Reply by Midnight-Knight** [reply]
https://nextjs.org/docs/app/api-reference/config/next-config-js/redirects

**Reply by dennzimm** [reply]
Thanks for the info :) let's see if that helps us too. I have only seen that it is probably due to `prefetch` from NextLink. If I set it to `false`, I no longer have these requests

**Reply by amir-rahmanii** [reply]
> Hi, at the moment, it has not been possible to completely fix the problem, but there is a feeling that we are moving in the right direction. It turned out that the `_rsc` query parameter is not available in any way for manipulation in `middleware.ts`, however, you can work with it in `next.config.js` . The following code allows you to configure traffic redirection if the `_rsc` query parameter is present in the URL.
> 
> ```
> module.exports = {
>   async redirects() {
>     return [
>       {
>         source: '/',
>         destination: '/',
>         has: [
>           {
>             type: 'query',
>             key: '_rsc',
>           }
>         ],
>         permanent: true,
>       },
>       {
>         source: '/:path*',
>         destination: '/:path*',
>         has: [
>           {
>             type: 'query',
>             key: '_rsc',
>           }
>         ],
>         permanent: true,
>       }
>     ]
>   },
>   ...
> }
> ```
> 
> As a result, from 1300+ original errors related to `_rsc` only 402 remain at the moment. Google started indexing some pages, but they are still not showing up in the search results, so the problem hasn't been fully resolved yet.
> 
> The version is next.js - 15.1.4

Is this still the recommended approach today?

**Comment by kstroka** [comment]
Hello there... do you have some solution for this? I have same issue, have thousants of ?rsc= in GSC ... what is best solution from SEO point of view for us?

**Comment by lightningspirit** [comment]
The best way to resolve this (did not work using `proxy.ts`) is to add the following parameter to the `headers()` property of yout `next.config.js` file:
```
{
    source: '/:path*',
    has: [
        {
        type: 'query',
        key: '_rsc',
        },
    ],
    headers: [
        {
        key: "X-Robots-Tag",
        value: "noindex, nofollow"
        }
    ],
}
```

Then will have a new header `X-Robots-Tag: noindex, nofollow` in the response of any request with `_rsc=` param query which will resolve the indexing problems.

**Comment by m-nathani** [comment]
Heads-up for anyone copying the `X-Robots-Tag` workaround from this thread — as written it also tags the **HTML document**, not just the RSC payload.

We shipped exactly this:

```js
{
  source: '/:path*',
  has: [{ type: 'query', key: '_rsc' }],
  headers: [{ key: 'X-Robots-Tag', value: 'noindex, nofollow' }],
}
```

and it was caught in review before it did damage. The reasoning error is that `_rsc` looks like it identifies an RSC request. It doesn't — it's only a client-side cache-buster (`NEXT_RSC_UNION_QUERY` in `app-router-headers.js`). The server decides RSC-ness from the `RSC` request header, never from the query.

And `matchHas` (`shared/lib/router/utils/prepare-destination.js`) evaluates only the conditions you list — for a `query` entry it reads `parsedUrl.query[key]` and never consults a header. Verified by calling it directly against `next@16.2.9`:

```
                                          query-only    query + RSC:1
crawler GET ?_rsc=abc, NO RSC header        MATCH           no
real prefetch ?_rsc=abc + RSC:1             MATCH          MATCH
clean page, no param                          no             no
```

So `GET /some/page?_rsc=abc` with **no** `RSC` header — what Googlebot sends once its renderer has queued the prefetch URL as a document, and what any third party can trigger by linking the param — is served the full HTML page and picks up `noindex, nofollow`.

That page carries its own `rel=canonical` to the clean URL. [Google's guidance](https://developers.google.com/search/docs/crawling-indexing/consolidate-duplicate-urls) is not to mix `noindex` with canonicalization, because a noindex inside a canonical group can propagate to the target — so the blast radius is the real pages, not the payload URLs. It's also consistent with what several people report in this thread: these URLs surface as *"Alternative page with proper canonical tag"*, which is only possible if Googlebot received HTML with a canonical tag rather than a `text/x-component` payload.

Adding the header condition fixes it:

```js
{
  source: '/:path*',
  has: [
    { type: 'query', key: '_rsc' },
    { type: 'header', key: 'RSC', value: '1' },
  ],
  headers: [{ key: 'X-Robots-Tag', value: 'noindex, nofollow' }],
}
```

`has` entries are AND-ed, so every genuine `text/x-component` response is still covered while the HTML case falls back to normal canonical consolidation — which is what you want for a real page plus a param.

Two notes on that version:

- `matchHas` requires a *truthy* value, so a valueless `?_rsc` matches nothing either way. Not reachable in practice, since every prefetch path sets a non-empty cache-busting hash.
- Worth being clear about what this buys: it stops those URLs being **indexed**, not **crawled**. Googlebot still has to fetch a URL to read a `noindex`. If your problem is crawl volume rather than the "Alternative page with proper canonical tag" reports, this rule won't move it.

**Comment by tpriyankaa** [comment]
This happens because Googlebot discovers and crawls the internal prefetch links generated by Next.js `<Link>` components, which append the `_rsc` query parameter to fetch raw React Server Component payloads. 

Since `robots.txt` blocks the entire page, it's not the right tool here. To stop Google Search Console from indexing these duplicate URLs while keeping your actual pages indexed, you should implement a dynamic canonical link configuration.

Here are the two ways to fix this:

### Option 1: Use Next.js Metadata API (Recommended)
Configure your root or layout metadata to dynamically strip query parameters from the canonical URL. This tells Google that `://domain.com` is identical to `://domain.com`, forcing Google to consolidate indexing onto the clean URL.

In your `app/layout.tsx` or specific page layout:
```ts
import { Metadata } from 'next';

export async function generateMetadata({ searchParams }): Promise<Metadata> {
  // Define your base production URL
  const baseUrl = 'https://yourdomain.com'; 
  
  return {
    alternates: {
      // This strips out internal query parameters like _rsc for Googlebot
      canonical: baseUrl, 
    },
  };
}
```

### Option 2: Clean the Canonical URL via Middleware
If you want a global solution that automatically drops `_rsc` or specific prefetch tokens across all canonical tags without modifying every page, you can intercept requests using `middleware.ts`:

```ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export function middleware(request: NextRequest) {
  const url = request.nextUrl.clone();
  
  // If a request comes in with the internal _rsc param, 
  // ensure the canonical header or context points to the clean URL
  if (url.searchParams.has('_rsc')) {
    url.searchParams.delete('_rsc');
    
    // Optional: Redirect search engines to the clean URL with a 301 
    // if it's an actual browser document request
    const response = NextResponse.next();
    response.headers.set('Link', `<${url.pathname}>; rel="canonical"`);
    return response;
  }
  
  return NextResponse.next();
}
```

### Option 3: Configure URL Parameters in Google Search Console
If you want to handle it directly inside GSC without modifying code:
1. Go to your **Google Search Console** dashboard.
2. Navigate to the legacy **URL Parameters** tool (if available for your property) or use the **Removals** tool.
3. Add `_rsc` as a parameter and set it to **"No: Doesn't affect page content (ex: tracks usage)"** to tell Googlebot to ignore it during indexing.
4.
